### PR TITLE
Retire FullNodeConfiguration

### PR DESCRIPTION
**Pull request by al-r3**
Remove all references to FullNodeConfiguration and just use NodeConfiguration everywhere.

### CODE REVIEW

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 1]
```diff
@@ -99,7 +98,7 @@ open class Node(configuration: FullNodeConfiguration,
     }
 
     override val log: Logger get() = logger
-    override val configuration get() = super.configuration as FullNodeConfiguration // Necessary to avoid init order NPE.
+    override val configuration get() = super.configuration  // Necessary to avoid init order NPE.
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: Is this overridden getter still needed?

File: node/src/main/kotlin/net/corda/node/internal/Node.kt [hunk 2]
```diff
@@ -168,13 +167,13 @@ open class Node(configuration: FullNodeConfiguration,
         return NodeMessagingClient(
                 configuration,
                 versionInfo,
-                serverAddress,
+                serverAddress!!,
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/Node.kt]: It looks like the `if (messagingServerAddress != null)` check above is not smart casting away the nullability. If you fix that then these two double-bangs can go away.

File: node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt [hunk 3]
```diff
@@ -68,7 +76,16 @@ data class BridgeConfiguration(val retryIntervalMs: Long,
 
 data class ActiveMqServerConfiguration(val bridge: BridgeConfiguration)
 
-data class FullNodeConfiguration(
+fun Config.parseAsNodeConfiguration(): NodeConfiguration {
```
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/config/NodeConfiguration.kt]: You can make this one line

File: node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationTest.kt [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+import net.corda.testing.ALICE
+import net.corda.testing.node.MockServices.Companion.makeTestDataSourceProperties
+import net.corda.testing.node.MockServices.Companion.makeTestDatabaseProperties
+import org.assertj.core.api.Assertions.assertThatThrownBy
+import org.junit.Test
+
+class NodeConfigurationTest {
+    @Test
+    fun `Artemis special characters not permitted in RPC usernames`() {
+        val map = mapOf<String, Any?>(
```
Comment (shamsasari) [inline node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationTest.kt]: This test can be done simply by creating a NodeConfigurationImpl - no need to use the typesafe config API. In fact there's a TODO in NodeConfigurationImpl.init  to move the username checks to ArtemisMessagingServer. There you could probably just need to deal with `User` objects for test.

File: node/src/main/kotlin/net/corda/node/internal/Node.kt (hunk 2 above)
Comment (al-r3) [reply node/src/main/kotlin/net/corda/node/internal/Node.kt]: fixed, and removed that funny construct with 'with' & friends.

File: node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationTest.kt (hunk 4 above)
Comment (al-r3) [reply node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationTest.kt]: I'd like NodeConfigurationImpl to be private, hence I don't want to use it in tests either.
Said that, I can't find any use of rpcUsers, unless I'm missing some reflection magic, there's an RPCUserService which gets somehow a list of Users, I'm not sure if that would be the place I should put that check in

Comment (shamsasari) [reply node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationTest.kt]: If it's too much then just create a NodeConfigurationImpl directly. You can always make it internal and annotate it visiblefortesting.

Comment (al-r3) [reply node/src/test/kotlin/net/corda/node/services/config/NodeConfigurationTest.kt]: found where it was used, (which was actually RPCUserServiceImpl). Moved all the logic and test logic there

File: node/src/main/kotlin/net/corda/node/internal/Node.kt (hunk 1 above)
Comment (al-r3) [reply node/src/main/kotlin/net/corda/node/internal/Node.kt]: removed (I was scared by the comment)

**Pull request merged**